### PR TITLE
Implement Trapped Rain Water problem using Dynamic Programming

### PR DESCRIPTION
**Pull request by TruongNhanNguyen**
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

### CODE REVIEW

File: src/test/java/com/thealgorithms/dynamicprogramming/TrappedRainWaterTest.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+public class TrappedRainWaterTest {
+    @ParameterizedTest
+    @MethodSource("testCases")
+    void testTrappedRainWater(int[] heights, int expected) {
+        assertEquals(expected, TrappedRainWater.trap(heights));
+    }
+
+    private static Stream<Arguments> testCases() {
+        return Stream.of(Arguments.of(new int[] {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}, 6), Arguments.of(new int[] {0, 2, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1}, 7), Arguments.of(new int[] {4, 2, 0, 3, 2, 5}, 9), Arguments.of(new int[] {3, 2, 1, 0, 1, 2, 3}, 9),
+            Arguments.of(new int[] {5, 4, 3, 2, 1, 2, 3, 4, 5}, 16), Arguments.of(null, 0), // Test case for null input
```
Comment (vil02) [inline src/test/java/com/thealgorithms/dynamicprogramming/TrappedRainWaterTest.java:19]: This comment does not add any value.

**Comment by vil02** [comment]
Since:
https://github.com/TheAlgorithms/Java/blob/295ac4b1ac2d43a4dea643380b6428a11c64500c/CONTRIBUTING.md?plain=1#L3
and the fact that [it is a leet code problem](https://leetcode.com/problems/trapping-rain-water/description/), I have to close it.

**Comment by TruongNhanNguyen** [comment]
> Since:
> 
> https://github.com/TheAlgorithms/Java/blob/295ac4b1ac2d43a4dea643380b6428a11c64500c/CONTRIBUTING.md?plain=1#L3
> 
> 
> and the fact that [it is a leet code problem](https://leetcode.com/problems/trapping-rain-water/description/), I have to close it.

_"Trapped rainwater" is a classic problem in computer science and algorithm design. It's often used as an example in dynamic programming, as well as in various coding challenges, including those found on platforms like LeetCode._

The problem involves calculating how much rainwater can be trapped in between blocks or walls of different heights. Solving this problem efficiently requires clever algorithmic thinking, and dynamic programming is one approach that can be used to solve it optimally.

So, while it's commonly encountered as a problem on coding platforms like LeetCode, you can see that it's not just a LeetCode problem; it's a broader algorithmic problem used to teach and practice dynamic programming techniques like other problem such as [coinchange](https://github.com/TheAlgorithms/Java/blob/master/src/main/java/com/thealgorithms/dynamicprogramming/CoinChange.java), [knapsack](https://github.com/TheAlgorithms/Java/blob/master/src/main/java/com/thealgorithms/dynamicprogramming/Knapsack.java) and so on.

Would you mind opening this PR again @vil02.

File: src/main/java/com/thealgorithms/dynamicprogramming/TrappedRainWater.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    /**
+     * Calculates the amount of trapped rainwater between blocks given their heights.
+     *
+     * @param height An array representing the heights of blocks.
+     * @return The total amount of trapped rainwater.
+     */
+    public static int trap(final int[] height) {
+        if (height == null || height.length == 0) return 0;
```
[the remark this answers is not in the document]
Comment (TruongNhanNguyen) [reply src/main/java/com/thealgorithms/dynamicprogramming/TrappedRainWater.java:17]: Yeah, I will update it.

File: src/test/java/com/thealgorithms/dynamicprogramming/TrappedRainWaterTest.java (hunk 1 above)
Comment (TruongNhanNguyen) [reply src/test/java/com/thealgorithms/dynamicprogramming/TrappedRainWaterTest.java:19]: I have removed this redundant comment.

**Pull request closed without merging**